src/ir.cpp

```diff
@@ -11434,8 +11434,10 @@ static TypeTableEntry *ir_analyze_bit_shift(IrAnalyze *ira, IrInstructionBinOp *
                 op1->value.type->data.integral.bit_count - 1);
 
         casted_op2 = ir_implicit_cast(ira, op2, shift_amt_type);
-        if (casted_op2 == ira->codegen->invalid_instruction)
+        if (casted_op2 == ira->codegen->invalid_instruction) {
+            ir_add_error(ira, &bin_op_instruction->base, buf_sprintf("RHS of shift is too large for LHS type"));
```

Comment:
To follow the pattern of the code around, this if check should be doing `if (type_is_invalid(casted_op2->value.type))`

In general, when you get an invalid type from an ir function, it means that an error has already been reported, and it would be inappropriate to report another error. I think what you want to do is add an error note to the original error. 

Consider the example:

```zig
const Foo = struct {};
var x: u8 = 1;
var y = x << Foo;
```

I haven't tried it but I believe this would emit `RHS of shift is too large for LHS type` which is not correct.

If we want to have this improved error message - which I agree with you that we do - then I think it would be better to look at the type of `shift_amt_type` before calling `ir_implicit_cast`. If it's an integer that is too big, then we can emit the specialized error message and return early. Otherwise, proceed with implicitly casting like normal.

Reply:
Right I was wondering about other cases that are not left shifts. I'll try that approach when I have a chance.